bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -1008,3 +1011,32 @@ func (r *KubeadmConfigReconciler) storeBootstrapData(ctx context.Context, scope
 	conditions.MarkTrue(scope.Config, bootstrapv1.DataSecretAvailableCondition)
 	return nil
 }
+
+// Ensure the bootstrap secret has the configOwner as a controller OwnerReference.
+func (r *KubeadmConfigReconciler) ensureBootstrapSecretOwnersRef(ctx context.Context, scope *Scope) error {
+	secret := &corev1.Secret{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: scope.Config.Namespace, Name: scope.Config.Name}, secret)
+	if err != nil {
+		// If the secret has not been created yet return early.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to add KubeadmConfig %s as ownerReference to bootstrap Secret %s", scope.ConfigOwner.GetName(), secret.GetName())
+	}
+	patchHelper, err := patch.NewHelper(secret, r.Client)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add KubeadmConfig %s as ownerReference to bootstrap Secret %s", scope.ConfigOwner.GetName(), secret.GetName())
+	}
+	secret.OwnerReferences = util.EnsureOwnerRef(secret.OwnerReferences, metav1.OwnerReference{
+		APIVersion: scope.ConfigOwner.GetAPIVersion(),
+		Kind:       scope.ConfigOwner.GetKind(),
+		UID:        scope.ConfigOwner.GetUID(),
+		Name:       scope.ConfigOwner.GetName(),
+		Controller: pointer.Bool(true),
```

Comment:
in my case, this operation occurs validation error such as..

    E1125 01:26:05.678896      12 controller.go:326] "Reconciler error" err="could not add KubeadmConfig hshong-test-control-plane-d5snx as ownerReference to bootstrap Secret hshong-test-control-plane-8rdj6: Secret \"hshong-test-control-plane-8rdj6\" is invalid: metadata.ownerReferences: Invalid value: []v1.OwnerReference{v1.OwnerReference{APIVersion:\"bootstrap.cluster.x-k8s.io/v1beta1\", Kind:\"KubeadmConfig\", Name:\"hshong-test-control-plane-8rdj6\", UID:\"6d1904bf-26b3-4edb-b7ca-b17854bd9b28\", Controller:(*bool)(0xc004564be0), BlockOwnerDeletion:(*bool)(nil)}, v1.OwnerReference{APIVersion:\"cluster.x-k8s.io/v1beta1\", Kind:\"Machine\", Name:\"hshong-test-control-plane-d5snx\", UID:\"7e3f24e8-8fe7-4d11-83d5-99dfe748cc97\", Controller:(*bool)(0xc004564bee), BlockOwnerDeletion:(*bool)(nil)}}: Only one reference can have Controller set to true. Found \"true\" in references for hshong-test-control-plane-8rdj6 and hshong-test-control-plane-d5snx" controller="kubeadmconfig" controllerGroup="bootstrap.cluster.x-k8s.io" controllerKind="KubeadmConfig" kubeadmConfig="cloudtech-team-dev/hshong-test-control-plane-8rdj6" namespace="cloudtech-team-dev" name="hshong-test-control-plane-8rdj6" reconcileID=3bb42394-0216-449b-ae0e-7baa7c1ac0d0

cause the machine resource of contol plane already takes `controller: true` attr for same KubeadmControlPlane resource.

> Multiple controllers can race to adopt a given object, but only one can win by being the first to add a ControllerRef to the object's OwnerReferences list. The losers will see their adoptions fail due to a validation error as explained above.
> 
> -   <https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md#adoption>

Reply:
Thx for the feedback @aanoaa !!

@killianmuldoon I"m a bit confused about the error message. We are adding the ConfigOwner as ownerRef, right? So we're essentially either adding a MachinePool or a Machine as owner. The error seems to suggest we add a KubeadmConfig as owner.

Independent of that. The error from @aanoaa seems to suggest that the KubeadmConfig is already owned by a a different Machine (probably?). 

Just for my understanding. We have e.g. the following owner chain:

1. Machine => (ownerRef) KubeadmConfig 
2. Machine => (ownerRef) Secret

This code established the second ownerRef and uses the first as data for that?

I assume we can be sure that 1. is always correct, can we then just drop additional wrong controller ownerRefs that might be already on the Secret?

Reply:
Edit: what I described seems to be indeed what our new code does, but looking at the Secret creation we should add the KubeadmConfig as owner here and not the configOwner (and then the error message is correct)

Reply:
Thanks for catching this @aanoaa  - I'm working on a fix now.

Reply:
Just to re-iterate.

Thank you very much @aanoaa for testing, finding and reporting this issue!!

Bug fixes should be merged in a few minutes across branches.